clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -211,50 +230,13 @@ public FetchRequestData build() {
                     log.debug("Built full fetch {} for node {} with {}.",
                               nextMetadata, node, partitionsToLogString(next.keySet()));
                 }
-                sessionPartitions = next;
                 next = null;
-                Map<TopicPartition, PartitionData> toSend =
-                    Collections.unmodifiableMap(new LinkedHashMap<>(sessionPartitions));
+                Map<TopicPartition, PartitionData> toSend = copySessionPartitions
+                    ? Collections.unmodifiableMap(new LinkedHashMap<>(sessionPartitions))
+                    : Collections.unmodifiableMap(sessionPartitions);
                 return new FetchRequestData(toSend, Collections.emptyList(), toSend, nextMetadata);
             }
 
-            List<TopicPartition> added = new ArrayList<>();
```

Comment:
Just a note that FetchSessionHandler is used by the consumer too, so we will either need two separate implementations or make sure it works just as well for the consumer too.

Reply:
One gotcha that I'm not sure whether we look out for yet. If the fetch session gets reset to a full fetch session https://github.com/apache/kafka/pull/7581/files#diff-38a1c48d28da7f7f671cef22421990e5L415, we should attempt to send the partitions in the order that they exist in the replica fetcher partitionStates. This ensures fairness if we are only ever performing full fetch sessions. Maybe this is not a problem in practice but we probably shouldn't change this behavior.

Reply:
Ack, I will try to see if consumer can leverage on the same path as well.

Reply:
Regarding `we should attempt to send the partitions in the order that they exist in the replica fetcher partitionStates.` Thinking about it a bit, I think the current logic actually still preserves it based on how we managed to update the list: on the very first initial fetch, the `sessionPartitions` would be set to `next` which is the same order of `fetchStates`, and then in future updates, when new partitions are added while looping over `fetchStates` they are added at the end of the sessionPartitions, and if they are modified they would be modified in place without changing the order.

WDYT?

Reply:
We initialize next to an empty hashmap on each pass, but we only add elements where updates are needed to it https://github.com/apache/kafka/pull/7581/files#diff-a8437f241a5ae585d5805ee769313080R263. Won't this result in only the updated partitions being sent in a full fetch request if the previous incremental response failed?

Reply:
In a full fetch request, we use the `sessionPartitions` directly instead of `next`, should that be sufficient?

Reply:
Ah, I missed that that was changed. Why do we still need next then? At a glance it seems to serve no purpose?

Reply:
We still need `next` in the actual incremental fetch, where a partition is neither `added` nor `removed` from the sessionPartitions, indicating that it is the same since last fetch, in this case it should not be included in `next`. I.e. we still need `next` to maintain "what's changed in this session" while `sessionPartitions` includes all the partitions we are interested in this session, no matter if they are changed or not.

Reply:
Oh, I see. next just means "updated or added" now. 

I'm still not sure sessionPartitions will necessarily match the order in partitionStates, as we are only ever updating it if partitions are updated (though usually we send those to the end of partition states too).